### ISSUE TITLE
Respect the lifetimebound in assignment operator.

### ISSUE THREAD

**Issue by hokein** [llvm/llvm-project#106372]
https://godbolt.org/z/43nf1v349

```
#include <vector>
#include <string>

template <typename T>
class  StatusOr {
public:
  template <typename U = T>
  StatusOr(U&& t [[clang::lifetimebound]]);

  template <typename U = T>
  StatusOr& operator=(U&& v [[clang::lifetimebound]]);
};

void test() {
  StatusOr<std::string_view> foo = std::string(); // #1 good, clang emits dangling diagnostic
  foo = std::string(); // #2, no dangling diagnostic
}
```

In the above example, clang doesn't catch the case 2, it would be nice to catch this case. We can assume that all assignment operators with a typical return type (`const reference&`) always return *this.
